File: packages/hubspot_flutter_chat/android/src/main/kotlin/io/github/marl0nl/hubspot_flutter_chat/HubspotChatPlugin.kt
package io.github.marl0nl.hubspot_flutter_chat

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.hubspot.mobilesdk.HubspotManager
import com.hubspot.mobilesdk.HubspotWebActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Android implementation of the hubspot_flutter chat bridge.
 *
 * It implements the Pigeon-generated [HubspotChatHostApi] and forwards each
 * call to HubSpot's official mobile chat SDK (`com.hubspot.mobilesdk`,
 * artifact `com.hubspot.mobilechatsdk:mobile-chat-sdk-android`). The SDK reads
 * its configuration from `assets/hubspot-info.json`; see the package README.
 *
 * Native → Dart chat events are delivered through [HubspotChatFlutterApi],
 * exposed to Dart as broadcast streams.
 */
class HubspotChatPlugin :
    FlutterPlugin,
    ActivityAware,
    HubspotChatHostApi {

    private lateinit var context: Context
    private var activity: Activity? = null
    private var defaultChatFlow: String? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    companion object {
        private const val CONFIG_ERROR = "hubspot_configuration_error"
        private const val NO_ACTIVITY = "no_activity"

        /** Set while an engine is attached so events can be sent to Dart. */
        @Volatile
        var flutterApi: HubspotChatFlutterApi? = null
            private set
    }

    // --- FlutterPlugin ---

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        HubspotChatHostApi.setUp(binding.binaryMessenger, this)
        flutterApi = HubspotChatFlutterApi(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HubspotChatHostApi.setUp(binding.binaryMessenger, null)
        flutterApi = null
        scope.cancel()
    }

    // --- ActivityAware (openChat needs a foreground Activity) ---

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    private fun manager(): HubspotManager = HubspotManager.getInstance(context)

    // --- HubspotChatHostApi ---

    override fun configure(setup: ChatSetupData, callback: (Result<Unit>) -> Unit) {
        defaultChatFlow = setup.defaultChatFlow
        try {
            // Loads assets/hubspot-info.json (portalId, hublet, environment...).
            manager().configure()
            callback(Result.success(Unit))
        } catch (e: Throwable) {
            callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
        }
    }

    override fun openChat(chatFlow: String?, callback: (Result<Unit>) -> Unit) {
        val host = activity
        if (host == null) {
            callback(
                Result.failure(
                    flutterError(NO_ACTIVITY, "No foreground Activity to present chat"),
                ),
            )
            return
        }
        // Pigeon's generated handler does not catch synchronous throws, so any
        // SDK/framework exception must be funnelled into the callback here (an
        // escaped exception would crash the app and hang the Dart future).
        try {
            val intent = Intent(host, HubspotWebActivity::class.java)
            (chatFlow ?: defaultChatFlow)?.let { intent.putExtra("chatflow", it) }
            host.startActivity(intent)
            flutterApi?.onChatOpened {}
            callback(Result.success(Unit))
        } catch (e: Throwable) {
            callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
        }
    }

    override fun closeChat(callback: (Result<Unit>) -> Unit) {
        // HubspotWebActivity is a full-screen Activity the user dismisses; the
        // SDK exposes no programmatic close, so this is a no-op success.
        callback(Result.success(Unit))
    }

    override fun setUserIdentity(
        identity: VisitorIdentity,
        callback: (Result<Unit>) -> Unit,
    ) {
        try {
            manager().setUserIdentity(identity.email, identity.identityToken)
            callback(Result.success(Unit))
        } catch (e: Throwable) {
            callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
        }
    }

    override fun setChatProperties(
        properties: Map<String, String>,
        callback: (Result<Unit>) -> Unit,
    ) {
        try {
            manager().setChatProperties(properties)
            callback(Result.success(Unit))
        } catch (e: Throwable) {
            callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
        }
    }

    override fun setPushToken(token: String, callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                manager().setPushToken(token)
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
            }
        }
    }

    override fun handlePushNotification(
        data: Map<String, String>,
        callback: (Result<Boolean>) -> Unit,
    ) {
        try {
            val isHubspot = HubspotManager.isHubspotNotification(data)
            if (isHubspot) flutterApi?.onNewMessage {}
            callback(Result.success(isHubspot))
        } catch (e: Throwable) {
            callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
        }
    }

    override fun logout(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                manager().logout()
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(flutterError(CONFIG_ERROR, e.message)))
            }
        }
    }

    private fun flutterError(code: String, message: String?): FlutterError =
        FlutterError(code, message, null)
}

File: packages/hubspot_flutter_chat/ios/hubspot_flutter_chat/Sources/hubspot_flutter_chat/HubspotChatPlugin.swift
import Flutter
import HubspotMobileSDK
import SwiftUI
import UIKit

/// Error surfaced when the chat UI cannot be presented.
enum HubspotChatError: Error {
    case noViewController
}

/// iOS implementation of the hubspot_flutter chat bridge.
///
/// Implements the Pigeon-generated `HubspotChatHostApi` protocol and forwards
/// each call to HubSpot's official iOS mobile chat SDK (`HubspotMobileSDK`). The
/// SDK reads its configuration from the app bundle (`Hubspot-Info.plist` /
/// config file); see the package README.
///
/// Native → Dart chat events are delivered through `HubspotChatFlutterApi`,
/// exposed to Dart as broadcast streams.
public class HubspotChatPlugin: NSObject, FlutterPlugin, HubspotChatHostApi {
    private var flutterApi: HubspotChatFlutterApi?
    private var defaultChatFlow: String?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = HubspotChatPlugin()
        HubspotChatHostApiSetup.setUp(
            binaryMessenger: registrar.messenger(),
            api: instance
        )
        instance.flutterApi = HubspotChatFlutterApi(
            binaryMessenger: registrar.messenger()
        )
    }

    // MARK: - HubspotChatHostApi

    func configure(
        setup: ChatSetupData,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        defaultChatFlow = setup.defaultChatFlow
        do {
            // Loads the bundled HubSpot configuration (portalId, hublet, ...).
            try HubspotManager.configure()
            completion(.success(()))
        } catch {
            completion(.failure(error))
        }
    }

    func openChat(
        chatFlow: String?,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        let flow = chatFlow ?? defaultChatFlow
        DispatchQueue.main.async {
            guard let top = Self.topViewController() else {
                completion(.failure(HubspotChatError.noViewController))
                return
            }
            let chatView = HubspotChatView(
                manager: HubspotManager.shared,
                chatFlow: flow
            )
            let host = UIHostingController(rootView: chatView)
            host.modalPresentationStyle = .fullScreen
            top.present(host, animated: true) {
                self.flutterApi?.onChatOpened { _ in }
            }
            completion(.success(()))
        }
    }

    func closeChat(completion: @escaping (Result<Void, Error>) -> Void) {
        DispatchQueue.main.async {
            let top = Self.topViewController()
            if top?.presentingViewController != nil {
                top?.dismiss(animated: true) {
                    self.flutterApi?.onChatClosed { _ in }
                }
            }
            completion(.success(()))
        }
    }

    func setUserIdentity(
        identity: VisitorIdentity,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        HubspotManager.shared.setUserIdentity(
            identityToken: identity.identityToken,
            email: identity.email
        )
        completion(.success(()))
    }

    func setChatProperties(
        properties: [String: String],
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        HubspotManager.shared.setChatProperties(data: properties)
        completion(.success(()))
    }

    func setPushToken(
        token: String,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        // Dart passes the APNs device token hex-encoded; the SDK wants raw Data.
        HubspotManager.shared.setPushToken(apnsPushToken: Self.dataFromHex(token))
        completion(.success(()))
    }

    func handlePushNotification(
        data: [String: String],
        completion: @escaping (Result<Bool, Error>) -> Void
    ) {
        // NEEDS ON-DEVICE / macOS VERIFICATION — cannot be compiled on Linux.
        //
        // Unlike Android (which authoritatively gates on
        // `HubspotManager.isHubspotNotification`), the iOS SDK 1.0.x exposes no
        // synchronous "is this push ours" check. We therefore return the
        // CONSERVATIVE value `false` and do NOT emit onNewMessage: returning
        // `true` unconditionally would make an app that suppresses its own
        // notification handling on a `true` result swallow *every* non-HubSpot
        // push. Handle HubSpot chat pushes on iOS via HubSpot's AppDelegate
        // integration instead. Wire this up (returning `true` only for genuine
        // HubSpot payloads) once the iOS SDK exposes a detection API.
        completion(.success(false))
    }

    func logout(completion: @escaping (Result<Void, Error>) -> Void) {
        HubspotManager.shared.clearUserData()
        completion(.success(()))
    }

    // MARK: - Helpers

    private static func dataFromHex(_ hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2, limitedBy: hex.endIndex)
                ?? hex.endIndex
            if let byte = UInt8(hex[index..<next], radix: 16) {
                data.append(byte)
            }
            index = next
        }
        return data
    }

    @MainActor
    private static func topViewController(
        _ base: UIViewController? = UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow }
            .first?.rootViewController
    ) -> UIViewController? {
        if let nav = base as? UINavigationController {
            return topViewController(nav.visibleViewController)
        }
        if let tab = base as? UITabBarController {
            return topViewController(tab.selectedViewController)
        }
        if let presented = base?.presentedViewController {
            return topViewController(presented)
        }
        return base
    }
}
